Monitor project build sources

eatkins opened the pull request:
In sbt 1.3.0, we only monitor build sources in the root project
directory and the root project meta build directory. This commit adds
these inputs for each project.

Fixes https://github.com/sbt/sbt/issues/5061.

eatkins committed:
Monitor project build sources

In sbt 1.3.0, we only monitor build sources in the root project
directory and the root project meta build directory. This commit adds
these inputs for each project.

Fixes https://github.com/sbt/sbt/issues/5061.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -371,13 +370,14 @@ object Defaults extends BuildCommon {
   )
 
-  def paths = Seq(
+  def paths: Seq[Setting[_]] = Seq(
```
```suggestion
  // included into JvmPlugin.projectSettings
  def paths: Seq[Setting[_]] = Seq(
```

eatkins replied:
Applied with slight rewording.

eed3si9n commented on main/src/main/scala/sbt/internal/nio/CheckBuildSources.scala:
```diff
@@ -59,7 +58,7 @@ private[sbt] object CheckBuildSources {
-      val baseDir = (LocalRootProject / baseDirectory).value
+      val baseDir = baseDirectory.value
       val sourceFilter = "*.{sbt,scala,java}"
       val projectDir = baseDir / "project"
```
Wouldn't this end up tracking `sub/project/*.scala`? Maybe metabuild source inputs needs to be split up, and added back at global level.

eatkins replied:
Yeah, that's intentional. In the project directory, top level scala files are allowed in compile.

eatkins replied:
It's common for builds to have something like project/Dependencies.scala and then refer to object Dependencies in the top level build.sbt.

eed3si9n replied:
I understand you need to track `project/*.scala`. There are two axes to traverse. One is subprojects, the other is metabuild. Currently those two concerns are mixed, so you end up with `sub/project/*.scala`.

eatkins replied:
Right, but why can't my subproject also have a metabuild?

eed3si9n replied:
A metabuild can have subprojects (, which I believe this code still doesn't track), but a metabuild of a subproject doesn't make sense since "subprojects" are members of the proper build, and only a build can have its meta-build.

eatkins replied:
I don't know the answer, but that appears to be true. I guess I misunderstood how recursion works with builds.

eed3si9n replied:
To put it into examples:

```
project/D.scala                // track
project/plugins.sbt            // track
project/foo/A.scala            // track (bonus points)
project/src/main/scala/B.scala // track
project/project/C.scala        // track
project/project/plugins.sbt    // track
build.sbt                      // track
sub/build.sbt                  // track
sub/project/E.scala            // do not track
```

eatkins replied:
I can't get the bonus points without degrading performance because searching for all of the scala files would require listing the target directory which can be extremely slow. Otherwise, I think that this PR now handles the other cases.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -371,13 +370,14 @@ object Defaults extends BuildCommon {
     sourceManaged := crossTarget.value / "src_managed",
-    resourceManaged := crossTarget.value / "resource_managed"
+    resourceManaged := crossTarget.value / "resource_managed",
+    Scope.Global / checkBuildSources / fileInputs ++= CheckBuildSources.buildSourceFileInputs.value,
```
```suggestion
    // lhs generates *.sbt files that needs to be tracked per subprojects
    Scope.Global / checkBuildSources / fileInputs ++= CheckBuildSources.buildSourceFileInputs.value,
```

eed3si9n reviewed (changes requested):
some comments.

eatkins committed:
Monitor meta build sources

We want to recursively monitor the project meta build, but we also want
to avoid listing directories that don't exist. To compromise, I rework
the buildSourceFileInputs to add the nested project directories if they
exist. Because the fileInputs are a setting, this means that adding a
new project directory and *.sbt or *.scala will not immediately trigger
a rebuild, but in most common cases, it will. I added a scripted test
for this.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -371,13 +371,16 @@ object Defaults extends BuildCommon {
-    resourceManaged := crossTarget.value / "resource_managed"
+    resourceManaged := crossTarget.value / "resource_managed",
+    // Adds subproject build.sbt files to the global list of build files to monitor
+    Scope.Global / checkBuildSources / fileInputs += baseDirectory.value.toGlob / "*.sbt"
```
👍

eed3si9n reviewed (dismissed):
LGTM. thanks @eatkins
